app/src/org/commcare/activities/HomeScreenBaseActivity.java

```diff
@@ -189,15 +194,86 @@ private void loadInstanceState(Bundle savedInstanceState) {
     }
 
     /**
-     * Set state that signifies activity was launch from external app.
+     * Set state that signifies activity was launch from external app
      */
     private void processFromExternalLaunch(Bundle savedInstanceState) {
         if (savedInstanceState == null && getIntent().hasExtra(DispatchActivity.WAS_EXTERNAL)) {
             wasExternal = true;
+            processSessionEndpoint();
             sessionNavigator.startNextSessionStep();
         }
     }
 
+    private void processSessionEndpoint() {
+        if (getIntent().hasExtra(SESSION_ENDPOINT_ID)) {
+            Endpoint endpoint = validateIntentForSessionEndpoint(getIntent());
+            if (endpoint != null) {
+                Vector<String> endpointArguments = endpoint.getArguments();
+                EvaluationContext evaluationContext = CommCareApplication.instance().getCurrentSessionWrapper().getEvaluationContext();
```

Comment:
Just realized: We should be doing the session stack reset here, rather than below, since the EC here depends on the current session state, and this could introduce unpredictable behavior for the stack ops depending on the session state when the callout was received.

Reply:
Another consideration on that point: We may need the Endpoint definition to be able to provide its own `<instance>` connectors, since it won't necessarily be connected to any of the other components where connectors are defined. This can be "future" addition, since most of the inbound expressions now will just be Strings or whatnot, but I'd expect us to run into it at some point.

Reply:
Sounds good. Keeping it as a future addition.